ERTP guide revision (#237)
Agoric/documentation: main/ertp/ERTP-Guide/PursesAndPayments.md

# Purses and Payments

> You can store `amount` objects in two kinds of objects:

Reviewer: No, you can store digital assets

- **[`purse`](https://agoric.com/documentation/glossary/#purse)**: Hold
  an `amount` of same branded digital assets until you withdraw part or
  all of them into a `payment` for use.
- **[`payment`](https://agoric.com/documentation/glossary/#purse)**:
  Hold an `amount` of digital assets to send to another party. Usually either
  a freshly created `amount` or one taken from a `purse`

Assets in purses and payments do not have to be currency-like, but can

> be anything you can construct a digital rights `amount` object for;

Reviewer: ```suggestion
be any kind of digital asset
```

swords to use in a game, rights to use a particular contract, theater
tickets, etc.

> Each non-empty `purse` and `payment` object contains exactly one
> `amount` object. While you may deposit or withdraw an `amount` to a

Reviewer: No, we really need to correct this. There is no such thing as an amount object. purses and payments hold digital assets of various amounts - could be empty, could be an amount that is non-empty. I think you are thinking of an amount as a thing that a purse or payment has or doesn't have. That's not right. It is a description of assets held. The description could be an empty amount (value = 0 for nat), or non-empty (value = 1 or more for nat).
```suggestion
Each non-empty `purse` and `payment` object contains exactly one
`amount` object. While you may deposit or withdraw an `amount` of digital assets to and from a
```

`purse`, it is repsectively added to the `purse` `amount`or split off into a `payment`

> object's `amount`. The same is true for adding to or splitting a

Reviewer: We cannot remove partial amounts from payments. Thus, nothing is ever split off from a payment in a deposit or withdraw.

Author: Guess I wasn't clear here. I was talking about taking value from a purse and putting it into a payment, such that you'd still have just one amount in the purse and one amount in the payment; you'd subtract what you took from the purse from its single thing, and add what to gave to the payment to the payment's single thing, rather than having two separate representations in both the purse and payment. Will rephrase.

`payment` object's `amount`.  Both a `purse` and a `payment` can only

> have an amount of a single `brand`.

Reviewer: I think this whole paragraph needs rewriting with the right concepts

With the exception of `mint` objects creating entirely new `payment`

> objects with newly created `amount` objects, a `payment` object is

Reviewer: No, there is no such thing as amount objects and they are not created when minted. The digital assets are created when minted, and the digital assets are of a certain amount.
```suggestion
objects with newly created `amount` objects, a `payment` object is
```

created by making a withdrawal on an `amount` from a `purse`. Note
that the `brand` of the new `payment` must be the same as the
`purse`'s associated `brand`.

Note that you don't transfer directly from `purse`s to

> another `purse`. Instead, to send and receive assets in ERTP you do these steps:

Reviewer: Should we talk about ERTP in the context of our larger system? I think so. In that case, we should talk about depositFacets and how to use them here. I can help with that if you have questions.

- Send assets:
  1. Withdraw an `amount`from a `purse`, creating a `payment`.
  2. Send this `payment` to a recipient object as a message.
- Receive assets:
  1. If you don't already have one, create a `purse` for the asset `brand`
     you'll receive. 

>   2. Get access to the asset `brand` you'll receive.

Reviewer: This step is unnecessary.

  3. Receive the message with the `payment` and deposit the `payment` in
     your `brand` appropriate `purse`. 

## Purses

> ![Purse methods](purse.svg)  

Reviewer: ```suggestion
![Purse methods](./purse.svg)  
```
No getIssuer method.
makeDepositFacet doesn't manipulate any digital assets.
Purses don't contain amounts, they contain amounts of digital assets.

Purses have five API methods:

> - [`purse.getIssuer()`](https://agoric.com/documentation/ertp/api/purse.html#purse-getissuer)

Reviewer: This isn't a method and shouldn't be for security purposes

Author: Done PR to remove from current API documentation.

  - Returns the `issuer` for the purse's `brand`. See the [Issuer](#issuers)
  section for more details about what this means and how it is used.
  - ```js
	const purseIssuer = purse.getIssuer();
	```
- [`purse.getCurrentAmount()`](https://agoric.com/documentation/ertp/api/purse.html#purse-getcurrentamount)

>   - Returns the current `amount` object stored in the purse. Note that

Reviewer: No, there is no such thing as an amount object. This method returns the `amount` of *digital assets* stored in the purse

    purses can be empty.
  - ```js
  const { issuer } = produceIssuer('bucks');
  const purse = issuer.makeEmptyPurse();
  const currentBalance = purse.getCurrentAmount();
  ```
- [`purse.deposit(payment, optAmount)`](https://agoric.com/documentation/ertp/api/purse.html#purse-deposit-payment-optamount)

>   - Deposit the `amount` contents of a `payment` into this purse,

Reviewer: no, deposit digital assets in payment into the purse

     const purse = issuer.makeEmptyPurse();
	 const payment = mint.mintPayment(amountMath.make(123));
	 const bucks123 = amountMath.make(123);
	 // Deposit a payment for 123 bucks into the purse. Ensure that this is the amount you expect.
	 purse.deposit(payment, bucks123);
	 const secondPayment = mint.mintPayment(amountMath.make(100));
	 // Throws error
	 purse.deposit(secondPayment, fungible123);
     ```
- [`purse.makeDepositFacet()`](https://agoric.com/documentation/ertp/api/purse.html#purse-makedepositfacet)
  - Creates a deposit-only facet on the `purse` that can be given
    to other parties. This lets them make a deposit, but not make
    withdrawals, to your purse.
  - ```js
     const depositOnlyFacet = purse.makeDepositFacet();
     // Give depositOnlyFacet to someone else. They can pass a payment
     // that will be deposited:
     depositOnlyFacet.receive(payment);
     ```
- [`purse.withdraw(amount)`](https://agoric.com/documentation/ertp/api/purse.html#purse-withdraw-amount)
  - Withdraw the `amount` from this purse into a new
    `payment`. Returns the new `payment`.

In addition, other objects, such as `issuer`, have methods which take

> purses as arguments or return purses. In particular:

Reviewer: Nothing has a purse argument

- [`issuer.makeEmptyPurse()`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-makeemptypurse)
  - Returns a new empty `purse` for storing an `amount` of the same `brand` as
  the `issuer`.
  - ```js
	   const { issuer, mint } = produceIssuer('bucks');
	   const purse = issuer.makeEmptyPurse();
   ```

## Payments

![Payment methods](payment.svg)   

Payments hold an `amount` of digital assets that are either about to
be sent to another party or which is in transit to another

> party. Payments are linear, meaning that either a `payment` has its full

Reviewer: Not sure what the distinction is between "about to be sent" or "in transit". In actuality, no digital assets are transferred until another party calls `claim(payment) or deposits the payment. 

Maybe let's say, payments hold an `amount` of digital assets which can be transferred to another party.

original balance, or it is used up entirely. It is impossible to
partially use a `payment`. In other words, if you create a `payment` with
an `amount` of 10 quatloos, the `payment` will always either have an

> `amount` of 10 quatloos or an `amount` of 0 quatloos, never anything

Reviewer: The payment will always have an `amount` of 10 quatloos or be deleted from the issuer.  It will never have an amount of 0 quatloos

in-between or larger than 10.

Payments can be deposited in purses, split into multiple 
payments, combined, and claimed (getting an exclusive payment and
revoking access from anyone else). 

Payments are often received from other actors and therefore should not
be trusted themselves. To get the balance of a payment, use the
`getAmountOf(payment)` method on the trusted `issuer` for the `brand`

> of the `payment`.  **tyg todo: How does one get said trusted issuer?**

Reviewer: To get the issuer for a brand that you didn't create, ask someone you trust. For example, the venue that is creating tickets for shows can be trusted to give you the issuer for the tickets. Or, a friend might have a cryptocurrency that they like and if you trust them, you might accept that the issuer they give you.

To convert a `payment` into a `purse`: 
1. Access a trusted `issuer` for the `payment`'s `brand`.. 
2. Create an empty purse with `issuer.makeEmptyPurse()`

> 3. Put the payment `amount` in the purse with `purse.deposit(payment)`

Reviewer: No, it should be 
```suggestion
3. Transfer the digital assets from the payment to the purse with `purse.deposit(payment)`
```

Payments have one API method, but many methods for other object
types have `payment` objects as arguments and effectively operate on `payments`.
- [`payment.getAllegedBrand()`](https://agoric.com/documentation/ertp/api/payment.html#payment-getallegedbrand)
  - Returns the `payment`'s `brand`, indicating what digital asset the
  payment purports to be. Note that `payment`s are not trusted, so this
  result should be treated with suspicion and verified before 

>   use. **tyg todo: How do I verify?**

Reviewer: Any issuer operations on the payment, if they succeed, verify the payment. For example, `E(issuer).isLive(payment)`

Other objects' `payment`-related methods:

> - [`issuer.getAmountOfPayment(payment)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-getamountof-payment)

Reviewer: ```suggestion
- [`issuer.getAmountOf(payment)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-getamountof-payment)
```

    true value. 
  ` ```js
	  const { issuer, mint, amountMath } = produceIssuer('bucks');
	  const payment = mint.mintPayment(amountMath.make(100));
	  issuer.getAmountOf(payment); // returns 100
    ```
- [`issuer.burn(payment, optAmount)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-burn-payment-optamount)
  - Burn all of the digital assets in the `payment`. `optAmount` is
    optional but if present, the `payment` balance must be equal to
    it. If `payment` is a promise, the operation proceeds after it resolves. 
  - ```js
	  const { issuer, mint, amountMath } = produceIssuer('bucks');
	  const amountToBurn = amountMath.make(10);
	  const paymentToBurn = mint.mintPayment(amountToBurn);
	  // burntAmount should equal 10
	  const burntAmount = issuer.burn(paymentToBurn, amountToBurn);
	  ```
- [`issuer.claim(payment, optAmount)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-claim-payment-optamount)
  - Transfer all assets from the `payment` to a returned new `payment`
    and burn the original. No other references to this payment survive, so 
    the new `payment`'s owner is the exclusive one. If `optAmount` is

>     present, `payment`'s balance is equal to `optAmount`'s, or it throsw

Reviewer: ```suggestion
    present, `payment`'s balance is equal to `optAmount`'s, or it throws
```
Could we have a spell check done before future reviews? I think that would save both of us a lot of time

    ```
- [`issuer.claim(payment, optAmount)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-claim-payment-optamount)
  - Transfer all assets from the `payment` to a returned new `payment`
    and burn the original. No other references to the argument `payment` survive, so 
    the new `payment` is the exclusive one. If `optAmount` is
    present, the `payment` balance must be equal to it or it throws
    an error. If `payment` is a promise, the operation proceeds after it resolves. 
  - ```js
    const { quatloosMint, quatloosIssuer, quatloosAmountMath } = makeIssuerKit('quatloos');
    const amountExpectedToTransfer = quatloosAmountMath.make(2);
    const originalPayment = quatloosMint.mintPayment(amountExpectedToTransfer);
    const newPayment = quatloosIssuer.claim(originalPayment, amountToTransfer);
    ```
- [`issuer.combine(paymentsArray)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-combine-paymentsarray)
  - Combine multiple `payment`s into one, returned, `payment`. If any `payment` in
  the array is a promise, the operation proceeds after every `payment`
  resolves. All `payments` in the array are burned.

>   - ```js

Author: ? I don't see the problem here. Maybe move it to before the preceding "promise" line? I added "upon successful completion" at the end, does that address it? Because from what I can tell, the original payments in that array of payments do all get burned in favor of the combined payment result.

    const { quatloosMint, quatloosIssuer, quatloosAmountMath } = makeIssuerKit('quatloos');
    // create an array of 100 payments of 1 unit each
    const payments = [];
    for (let i = 0; i < 100; i += 1) {
      payments.push(quatloosMint.mintPayment(quatloosAmountMath.make(1)));
    }
    // combinedPayment equals 100
    const combinedPayment = quatloosIssuer.combine(payments);
    ```
- [`issuer.split(payment, paymentAmountA)`](https://agoric.com/documentation/ertp/api/issuer.html#issuer-split-payment-paymentamounta)

>   - Split one `payment` into two new ones, A and B, returned in

Reviewer 2: if paymentAmountA is larger than payment's amount, throws.
